### PR TITLE
create isUnaExp(), isBinExp(), isBinAssignExp()

### PR DESCRIPTION
**Pull request by WalterBright**
Use flags instead of inheritance/casting to determine these categories.

### CODE REVIEW

File: src/dmd/expression.d [hunk 1]
```diff
@@ -1742,11 +1742,21 @@ extern (C++) abstract class Expression : ASTNode
         inout(PrettyFuncInitExp) isPrettyFuncInitExp() { return op == EXP.prettyFunction ? cast(typeof(return))this : null; }
         inout(ClassReferenceExp) isClassReferenceExp() { return op == EXP.classReference ? cast(typeof(return))this : null; }
         inout(ThrownExceptionExp) isThrownExceptionExp() { return op == EXP.thrownException ? cast(typeof(return))this : null; }
-    }
 
-    inout(BinAssignExp) isBinAssignExp() pure inout nothrow @nogc
-    {
-        return null;
+        inout(UnaExp) isUnaExp() pure inout nothrow @nogc
```
Comment (thewilsonator) [inline src/dmd/expression.d:1746]: if these are no longer overridden, props make them final?

Comment (WalterBright) [reply src/dmd/expression.d:1746]: See line 1622

**Comment by WalterBright** [comment]
dscanner is complaining about:
```
/var/lib/buildkite-agent/builds/ci-agent-59bf1ce0-2222-401d-afec-f34be92c8925-4/dlang/dmd/dmd/src/dmd/expression.d(7086:14)[warn]: This enum may lead to unnecessary allocation at run-time. Use 'static immutable Eunary = [ ...' instead.
```
but backend/oper.d does the same thing. How do I disable that style check?

**Comment by WalterBright** [comment]
I found it. Never mind.

**Review by thewilsonator — APPROVED** [review]

**Comment by ibuclaw** [comment]
This PR introduced a regression https://issues.dlang.org/show_bug.cgi?id=22788

**Pull request merged**